src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSLambdaHttpUtils.cs

```diff
@@ -0,0 +1,134 @@
+            {
+                case APIGatewayProxyRequest request:
+                    httpScheme = AWSLambdaUtils.GetHeaderValues(request, HeaderXForwardedProto)?.FirstOrDefault();
+                    httpTarget = string.Concat(request.Path ?? string.Empty, ConstructQueryString(request));
+                    httpMethod = request.HttpMethod;
+                    var hostHeader = AWSLambdaUtils.GetHeaderValues(request, HeaderHost)?.FirstOrDefault();
+                    (hostName, hostPort) = GetHostAndPort(httpScheme, hostHeader);
+                    break;
+                case APIGatewayHttpApiV2ProxyRequest requestV2:
+                    httpScheme = AWSLambdaUtils.GetHeaderValues(requestV2, HeaderXForwardedProto)?.FirstOrDefault();
+                    httpTarget = string.Concat(requestV2?.RawPath ?? string.Empty, ConstructQueryString(requestV2));
+                    httpMethod = requestV2?.RequestContext?.Http?.Method;
+                    var hostHeaderV2 = AWSLambdaUtils.GetHeaderValues(requestV2, HeaderHost)?.FirstOrDefault();
+                    (hostName, hostPort) = GetHostAndPort(httpScheme, hostHeaderV2);
+                    break;
+            }
+
+            tags.AddTagIfNotNull(SemanticConventions.AttributeHttpScheme, httpScheme);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeHttpTarget, httpTarget);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeHttpMethod, httpMethod);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeNetHostName, hostName);
+            tags.AddTagIfNotNull(SemanticConventions.AttributeNetHostPort, hostPort);
+
+            return tags;
+        }
+
+        internal static void SetHttpTagsFromResult(Activity activity, object result)
+        {
+            switch (result)
+            {
+                case APIGatewayProxyResponse response:
+                    activity.SetTag(SemanticConventions.AttributeHttpStatusCode, response.StatusCode);
+                    break;
+                case APIGatewayHttpApiV2ProxyResponse responseV2:
+                    activity.SetTag(SemanticConventions.AttributeHttpStatusCode, responseV2.StatusCode);
+                    break;
+            }
+        }
+
+        internal static string ConstructQueryString(APIGatewayProxyRequest request)
+        {
+            if (request.MultiValueQueryStringParameters == null)
+            {
+                return string.Empty;
+            }
+
+            var items = new List<string>();
+            foreach (var name in request.MultiValueQueryStringParameters.Keys)
+            {
+                // Multiple values for the same parameter will be added to query
+                // as ampersand separated: name=value1&name=value2
+                foreach (var value in request.MultiValueQueryStringParameters[name])
+                {
+                    items.Add(string.Concat(name, "=", HttpUtility.UrlEncode(value)));
+                }
+            }
```

Comment:
This nested foreach looks like it could be a SelectMany. Although performance might suffer, so it's up to you.

```suggestion
            var items = request.MultiValueQueryStringParameters.SelectMany(kv => kv.Value.Select($"{name}={HttpUtility.UrlEncode(value)});
```

If you keep using a foreach, you should avoid the `[]`-lookup and instead iterate key-value-pairs in the outer loop.

(also consider the previous comment about encoding the key/name which I did not address here)

Reply:
I would prefer to keep the nested loop as not sure about performance of  `SelectMany`. Made another optimisation: got rid of the List<T> and used StringBuilder instead.